Allow extension pages to set frame-specific website policies

charliewolfe opened the pull request:
#### 8fa47d3ff3db03b72ba4636a47b2da28e3dffb6d
<pre>
Allow extension pages to set frame-specific website policies
<a href="https://bugs.webkit.org/show_bug.cgi?id=255880">https://bugs.webkit.org/show_bug.cgi?id=255880</a>
rdar://104453529

Reviewed by Alex Christensen.

This patch makes changes to allow extensions pages to have subframes which have website
policies that are different from the main frame. For now, this only applies to the policies
allowPrivacyProxy, networkConnectionIntegrityPolicy, and contentExtensionEnablement.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::urlForBindings const):
(WebCore::Document::referrerForBindings):
* Source/WebCore/page/UserContentProvider.cpp:
(WebCore::ruleListFilter):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::networkConnectionIntegrityPolicy const):
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::addParametersShared):
* Source/WebKit/WebProcess/Network/WebSocketChannel.cpp:
(WebKit::WebSocketChannel::connect):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsitePolicies.mm:

Canonical link: <a href="https://commits.webkit.org/263442@main">https://commits.webkit.org/263442@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4acb8f824d75688848fe67a2d6710e1d1b8aa5bd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4597 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4717 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4870 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6090 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4754 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4588 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4856 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4676 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4993 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4666 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4759 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4119 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6100 "Built successfully") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4111 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/9109 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4107 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4185 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5735 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4575 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3727 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4104 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1138 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8147 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4465 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3578,8 +3578,11 @@ const URL& Document::urlForBindings() const
-        auto* topDocumentLoader = topDocument().loader();
-        if (!topDocumentLoader || !topDocumentLoader->networkConnectionIntegrityPolicy().contains(NetworkConnectionIntegrity::Enabled))
+        auto* targetLoader = topDocument().loader();
+        if (targetLoader && LegacySchemeRegistry::isUserExtensionScheme(targetLoader->request().url().protocol()) && !LegacySchemeRegistry::isUserExtensionScheme(url().protocol()))
```
We should not use LegacySchemeRegistry::isUserExtensionScheme, and we should be working to remove it.  Use URL::protocolIsInHTTPFamily instead.

charliewolfe replied:
Changed this to use URL::hasSpecialScheme().

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsitePolicies.mm:
```diff
@@ -1833,3 +1835,148 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
+
+TEST(WebpagePreferences, ExtensionPageContentBlockers)
+{
+    ClassMethodSwizzler swizzler([WKWebView class], @selector(handlesURLScheme:), reinterpret_cast<IMP>(overrideHandlesURLScheme));
```
Don't do this.

charliewolfe replied:
Removed.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsitePolicies.mm:
```diff
@@ -1833,3 +1835,148 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
+    }];
+
+    auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
+    [configuration setURLSchemeHandler:handler.get() forURLScheme:@"safari-extension"];
```
Let's not use this scheme in the tests.  WebKit should know nothing about this scheme.

charliewolfe replied:
Changed the scheme to `test`.

achristensen07 commented on Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
```diff
@@ -327,7 +327,11 @@ static void addParametersShared(const LocalFrame* frame, NetworkResourceLoadPara
 
-    parameters.allowPrivacyProxy = mainFrameDocumentLoader ? mainFrameDocumentLoader->allowPrivacyProxy() : true;
+    auto* targetDocumentLoader = mainFrameDocumentLoader;
+    if (!targetDocumentLoader->request().url().hasSpecialScheme() && frame->document()->url().hasSpecialScheme())
```
It looks like this needs a null check like it had before this change.

achristensen07 commented on Source/WebKit/WebProcess/Network/WebSocketChannel.cpp:
```diff
@@ -137,8 +137,11 @@ WebSocketChannel::ConnectStatus WebSocketChannel::connect(const URL& url, const
+            auto* targetDocumentLoader = mainFrameDocumentLoader;
+            if (!targetDocumentLoader->request().url().hasSpecialScheme() && frame->document()->url().hasSpecialScheme())
+                targetDocumentLoader = frame->document()->loader();
+            allowPrivacyProxy = targetDocumentLoader->allowPrivacyProxy();
```
need another null check here, because targetDocumentLoader may have changed.

achristensen07 commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3578,8 +3578,11 @@ const URL& Document::urlForBindings() const
-        auto* topDocumentLoader = topDocument().loader();
-        if (!topDocumentLoader || !topDocumentLoader->networkConnectionIntegrityPolicy().contains(NetworkConnectionIntegrity::Enabled))
+        auto* targetLoader = topDocument().loader();
+        if (targetLoader && !targetLoader->request().url().hasSpecialScheme() && url().hasSpecialScheme())
```
The top document loader should probably check hasSpecialScheme, but this document's URL should probably just check protocolIsInHTTPFamily.  I don't think we want to do this for file URLs loaded in iframes, for example.

Also, I think looking at a new document loader needs a linked-on-or-after check.  Otherwise, existing applications that expect the top document's document loader to be consulted without giving the same policies to iframes will be broken.

achristensen07 commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3578,8 +3578,11 @@ const URL& Document::urlForBindings() const
 
-        auto* topDocumentLoader = topDocument().loader();
-        if (!topDocumentLoader || !topDocumentLoader->networkConnectionIntegrityPolicy().contains(NetworkConnectionIntegrity::Enabled))
+        auto* targetLoader = topDocument().loader();
```
I'm not sure "target" a good word here.  "policyDocumentLoader" also has other meaning.  What about "policySource" or "policyContainer"?  Same elsewhere.

charliewolfe replied:
Changed this to policySource.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsitePolicies.mm:
```diff
@@ -1833,3 +1834,146 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
+
+    auto request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:@"test:///index.html"]];
+    [webView loadRequest:request];
+    [navigationDelegate waitForDidFinishNavigation];
```
This doesn't wait for the postMessage to go through, so this test will likely be flaky.  Using something like _test_waitForAlert and alerting the results instead will fix this.

charliewolfe replied:
The postMessage should never fire in this case, since the script should be blocked. Instead, I added an additional 0.5 second wait. The other tests should be able to do this though, so I changed those to wait for an alert.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsitePolicies.mm:
```diff
@@ -1833,3 +1834,146 @@ - (void)webView:(WKWebView *)webView runJavaScriptAlertPanelWithMessage:(NSStrin
+    EXPECT_WK_STREQ("test:///index.html", results[0]);
+
+    [[WKContentRuleListStore defaultStore] _removeAllContentRuleLists];
+}
```
Could you also add a test verifying that if you load an http page with an http iframe, or a custom scheme page with a custom scheme iframe, the main frame policies still apply everywhere?

charliewolfe replied:
Sure, added.

charliewolfe committed:
Allow extension pages to set frame-specific website policies
https://bugs.webkit.org/show_bug.cgi?id=255880
rdar://104453529

Reviewed by Alex Christensen.

This patch makes changes to allow extensions pages to have subframes which have website
policies that are different from the main frame. For now, this only applies to the policies
allowPrivacyProxy, networkConnectionIntegrityPolicy, and contentExtensionEnablement.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::urlForBindings const):
(WebCore::Document::referrerForBindings):
* Source/WebCore/page/UserContentProvider.cpp:
(WebCore::ruleListFilter):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::networkConnectionIntegrityPolicy const):
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::addParametersShared):
* Source/WebKit/WebProcess/Network/WebSocketChannel.cpp:
(WebKit::WebSocketChannel::connect):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsitePolicies.mm:

Canonical link: https://commits.webkit.org/263442@main
